README.md
# CowBulls Activity
Sugar Activity of the classic "CowBulls" game, with a slight twist ;)

## About:
 - This is a [sugar](https://sugarlabs.org/) activity
 - After each round, your score increases by the `number of lives remaining + 1`

## Legend:
 - Bull : corresponding digit is at its correct place.
 - Cow  : corresponding digit is in the number but not at its correct place
 - Cross: corresponding digit is not present in th actual number

## Known Bugs:
 - Numpad keys 3, 5 and 0 aren't detected.<br>
   Log: `Key KP_Next unrecognised`

## Credits:
 - Icons borrowed from multiple authors via [FlatIcon](https://www.flaticon.com)

## Screenshots:
 - Round in progress (Level: Easy)
 ![Round in progress](screenshots/en/1.jpg)
 - Round won (Level: Easy)
 ![Round won](screenshots/en/2.jpg)


NEWS
3
* Regenerate POT file (Jui Pradhan)
* Fix port to python3 regression and logical error (Jui Pradhan),
* Fix TypeError (Jui Pradhan),
* Fix: won/lost emoji is not centered (Jui Pradhan),
* Fix EOL for new-game.svg (Jui Pradhan),
* Flake8 fixes (Jui Pradhan),
* Fix level drop-down (Jui Pradhan),
* Fix Hard level number entry to fit the screen (Jui Pradhan),
* Handle videoresize (Ayush Nawal),
* Add icon for restart (Ayush Nawal),
* Fix force redraw (Ayush Nawal, Jui Pradhan),
* Remove black screen when play button is pressed (Ayush Nawal),
* Port to python3 (Jui Pradhan, Ayush Nawal),
* Improve Readme and instructions (Rahul Bothra),
* Disable next round button on changing levels (Rahul Bothra)

2
* Fix image overlap on different screen sizes

1.1
* Translucent screen on round ending

1
* Added Initial Version
* Contains three levels:
*  - Easy (3 Numbers)
*  - Medium (4 Numbers),
*  - Hard (5 Numbers)


g.py
import os
import pygame

app = 'CowBulls'
BG_COLOR = (220, 220, 220)
images = {}


def init():
    global redraw, screen, images
    global scale, w, h, pointer, pos
    global INPUT_SIZE, DISP_SIZE, DIALPAD
    global ATTEMPT, XGAP, DKP

    redraw = True
    for root, subdirs, files in os.walk('data'):
        for file in files:
            load_image(os.path.join(root, file))

    screen = pygame.display.get_surface()
    pygame.display.set_caption(app)
    pygame.display.flip()
    w, h = screen.get_size()
    scale = min([w / 1200.0, h / 900.0])
    X_INIT = (w - 1200 * scale) / 2
    Y_INIT = (h - 900 * scale) / 2

    pos = pygame.mouse.get_pos()
    pointer = images['data/pointer.png']
    pygame.mouse.set_visible(False)
    MARGIN = 10 * scale
    INPUT_SIZE = 75 * scale
    DISP_SIZE = 55 * scale
    XGAP = MARGIN + INPUT_SIZE
    DKP = MARGIN + DISP_SIZE

    ATTEMPT = (X_INIT + 60 * scale, Y_INIT + 25 * scale)
    DIALPAD = (X_INIT + 900 * scale, Y_INIT + 200 * scale)


def load_image(file_path):
    img = pygame.image.load(file_path)
    images[file_path] = img


def clear_patch(pos, size=0):
    if not size:
        pygame.draw.rect(
            screen,
            BG_COLOR,
            (pos[0],
             pos[1],
                INPUT_SIZE,
                INPUT_SIZE))
    else:
        pygame.draw.rect(screen, BG_COLOR, (pos[0], pos[1], size, size))


main.py
import pygame
import sys
from pygame.locals import *
import utils
import g

X = 800
Y = 200
XX = 60
YY = 25
MARGIN = 10
INPUT_SIZE = 70
DISP_SIZE = 55
XGAP = MARGIN + INPUT_SIZE
DPX = MARGIN + DISP_SIZE
BG_COLOR = (220, 220, 220)


class CowBulls:
    def __init__(self, level=3):
        self.attempts = 0
        self.num = [int(x) for x in str(utils.get_random(level))]
        self.level = level
        self.game_over = False
        self.lives, self.offset = utils.get_lives(self.level)
        self.input = []
        self.highlight = 0

    def display(self):
        for x in range(3):
            for y in range(4):
                utils.blit_offset('dialpad/' + str(1 + x + 3 * y), (X, Y), (x, y))

        for x in range(int(self.level)):
            utils.blit_offset('entries/entry-small', (XX, YY + DISP_SIZE), (x, 0), 1)
            utils.blit_offset('status/bull', (XX, YY), (self.level + .5 + x, 0), 1)
            utils.blit_offset('entries/question', (XX, YY), (x, 0), 1)

        for x in range(self.level):
            utils.blit_offset('dialpad/entry-big', (X, Y), (x + self.offset, -0.7))
        utils.blit_offset('lives/lives', (X, Y), (0, 5))
        utils.blit_offset('lives/' + str(self.lives), (X, Y), (1, 5))

    def put_num(self, valy):
        if len(self.input) < self.level:
            utils.blit_offset('dialpad/' + str(valy), (X, Y), (self.offset + len(self.input), -1.6))
            self.input.append(int(valy))

    def clear(self):
        for x in range(self.level):
            self.delete()

    def delete(self):
        if (len(self.input) != 0):
            self.clear_patch(
                (X + (self.offset + len(self.input) - 1) * (XGAP), Y - 1.6 * XGAP))
            self.input.pop()

    def do_button(self, pos):
        pos = ((pos[0] - X) / XGAP), ((pos[1] - Y) / XGAP)
        num = 3 * pos[1] + pos[0] + 1
        if num > 0 and num <= 9:
            self.put_num(num)
        elif num == 10:
            self.enter()
        elif num == 11:
            self.put_num(0)
        elif num == 12:
            self.delete()

    def clear_patch(self, pos, size=INPUT_SIZE):
        pygame.draw.rect(g.screen, BG_COLOR, (pos[0], pos[1], size, size))

    def remove_glow(self):
        if self.highlight:
            if self.highlight % 3:
                x, y = (self.highlight % 3) - 1, self.highlight / 3
            else:
                x, y = 2, self.highlight / 3 - 1
            self.clear_patch((X + x * (XGAP), Y + y * XGAP))
            utils.blit_offset('dialpad/' + str(self.highlight), (X, Y), (x, y))
            self.highlight = 0

    def glow(self, num, pos):
        utils.blit_offset('dialpad/highlight', (X, Y), pos)
        utils.blit_offset('dialpad/' + str(num), (X, Y), pos)

    def highlight_bt(self, pos):
        self.remove_glow()
        pos = ((pos[0] - X) / XGAP), ((pos[1] - Y) / XGAP)

        if pos[0] in range(3) and pos[1] in range(4):
            num = 3 * pos[1] + pos[0] + 1
            self.glow(num, pos)
            self.highlight = num
        else:
            self.highlight = 0

    def attempt_disp(self):
        for x in range(self.level):
            if self.num[x] == self.input[x]:
                status = 'bull'
            elif self.input[x] in self.num:
                status = 'cow'
            else:
                status = 'cancel'
            utils.blit_offset('status/' + status, (XX, YY), (self.level + .5 + x, self.attempts + 1), 1)
            utils.blit_offset('entries/' + str(self.input[x]), (XX, YY), (x, self.attempts + 1), 1)

    def display_answer(self):
        for x in range(self.level):
            tt = self.lives + 2
            utils.blit_offset('entries/' + str(self.num[x]), (XX, YY), (x, self.lives + 2), 1)
            utils.blit_offset('status/bull', (XX, YY), (x + 0.5 + self.level, self.lives + 2), 1)

    def enter(self):
        if len(self.input) != self.level:
            pass
        else:
            self.attempt_disp()
            self.attempts += 1
            utils.blit_offset('lives/' + str(self.lives - self.attempts)), (X, Y), (1, 5))
            status = None
            if self.input == self.num:
                status = 'won'
            elif self.attempts == self.lives:
                self.display_answer()
                status = 'lost'
            if status:
                utils.load_blit(status, (550, 300))
                self.game_over = True

            self.clear()

    def run(self):
        g.init()
        self.display()
        while True:
            for event in pygame.event.get():
                if not self.game_over:
                    if event.type == pygame.MOUSEBUTTONDOWN:
                        self.do_button(event.pos)
                    elif event.type == pygame.MOUSEMOTION:
                        self.highlight_bt(event.pos)
                    elif event.type == pygame.KEYDOWN:
                        value = utils.get_input(event.key)
                        if value:
                            self.put_num(value)
                        elif event.key == pygame.K_BACKSPACE:
                            self.delete()
                        elif event.key in [pygame.K_RETURN, pygame.K_KP_ENTER]:
                            self.enter()
            pygame.display.update()

    def restart(self):
        self.init(self.level)
        self.run()

    def change_level(self, level):
        self.init(level)
        self.run()



if __name__ == "__main__":
    pygame.init()
    pygame.display.set_mode((1200, 700))
    game = CowBulls(int(sys.argv[1]))
    game.run()
    pygame.display.quit()
    pygame.quit()
    sys.exit(0)


utils.py
# utils.py
"""
    Copyright (C) 2018  Rahul Bothra

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

"""

import pygame
import g
import random

Numbers = {
    pygame.K_1: '1',
    pygame.K_2: '2',
    pygame.K_3: '3',
    pygame.K_4: '4',
    pygame.K_5: '5',
    pygame.K_6: '6',
    pygame.K_7: '7',
    pygame.K_8: '8',
    pygame.K_9: '9',
    pygame.K_0: '0',
    pygame.K_KP0: '0',
    pygame.K_KP1: '1',
    pygame.K_KP2: '2',
    pygame.K_KP3: '3',
    pygame.K_KP4: '4',
    pygame.K_KP5: '5',
    pygame.K_KP6: '6',
    pygame.K_KP7: '7',
    pygame.K_KP8: '8',
    pygame.K_KP9: '9',
}

levels = {
    3: (5, 0),
    4: (6, -0.5),
    5: (7, -1)
}


def get_num(level):
    num = [0]
    while num[0] == 0:
        if level == 3:
            num = random.sample(range(0, 10), 3)
        else:
            num = [random.randint(0, 10) for _ in range(level)]
    return num


def get_input(key):
    if key in Numbers:
        return Numbers[key]
    else:
        return None


def blit_offset(file_path, pos, offset, flag=0):
    if not flag:
        load_blit(file_path,
                  (pos[0] + offset[0] * g.XGAP, pos[1] + offset[1] * g.XGAP))
    if flag:
        load_blit(file_path,
                  (pos[0] + offset[0] * g.DKP, pos[1] + offset[1] * g.DKP))


def load_blit(file_path, pos):
    file_path = 'data/' + file_path + '.png'
    sizex, sizey = g.images[file_path].get_rect().size
    img = pygame.transform.smoothscale(
        g.images[file_path], (int(sizex * g.scale), int(sizey * g.scale)))
    if file_path == 'data/won.png' or file_path == 'data/lost.png':
        width = img.get_width()
        g.screen.blit(img, (pos[0] - width / 2, pos[1]))
    else:
        g.screen.blit(img, pos)


def get_lives(level):
    if level in levels:
        return levels[level]
    else:
        return (0, 0)


CowBulls.py
#!/usr/bin/python
# CowBulls.py

"""
    Copyright (C) 2018  Rahul Bothra

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.
"""
"""
    Acknowledgement:
    Icons borrowed from multiple authors via FlatIcon: https://www.flaticon.com
"""

import sys
import pygame

import gi
gi.require_version('Gtk', '3.0')
from gi.repository import Gtk

from sugar3.graphics.style import GRID_CELL_SIZE

import g
import utils


class CowBulls:
    def __init__(self, level=3, parent=None):
        self.parent = parent
        self.level = level
        self.journal = True  # set to False if we come in via main()
        self.num = utils.get_num(self.level)
        self.lives, self.offset = utils.get_lives(self.level)
        self.status = None
        self.input = []
        self.compare_list = []
        self.attempts_list = []
        self.score = 0
        self.parent.update_score(self.score)
        self.going = True
        self.clock = pygame.time.Clock()

    def display(self):
        g.screen.fill(g.BG_COLOR)
        for x in range(3):
            for y in range(4):
                utils.blit_offset('dialpad/no-glow',
                                  g.DIALPAD, (x, y))
                utils.blit_offset('dialpad/' + str(1 + x + 3 * y),
                                  g.DIALPAD, (x, y))
        for x in range(int(self.level)):
            utils.blit_offset('entries/entry-small',
                              (g.ATTEMPT[0], g.ATTEMPT[1] + g.DISP_SIZE),
                              (x, 0), 1)
            utils.blit_offset('status/bull', g.ATTEMPT,
                              (self.level + .5 + x, 0), 1)
            utils.blit_offset('entries/question', g.ATTEMPT, (x, 0), 1)
            utils.blit_offset('dialpad/entry-big', g.DIALPAD,
                              (x + self.offset, -0.7))

        for x in range(len(self.input)):
            utils.blit_offset('dialpad/no-glow', g.DIALPAD,
                              (self.offset + x, -1.6))
            utils.blit_offset('dialpad/' + str(self.input[x]), g.DIALPAD,
                              (self.offset + x, -1.6))

        utils.blit_offset('lives/lives', g.DIALPAD, (0, 4.5))
        utils.blit_offset('lives/' + str(self.lives -
                          len(self.attempts_list)), g.DIALPAD, (1, 4.5))
        self.highlight()
        self.result()
        self.attempt_disp()
        if self.status == 'lost':
            self.display_answer()

    def result(self):
        if self.status:
            self.alphasurface = pygame.Surface((g.w, g.h))
            self.alphasurface.fill((100, 100, 100))
            self.alphasurface.set_alpha(200)
            g.screen.blit(self.alphasurface, pygame.Rect(0, 0, g.w, g.h))
            utils.load_blit(self.status, (g.w / 2, g.h / 2 - (128 * g.scale)))

    def delete(self):
        if (len(self.input) != 0):
            self.input.pop()

    def g_init(self):
        g.init()

    def do_key(self, event):
        value = utils.get_input(event.key)
        if value:
            if len(self.input) < self.level:
                self.input.append(int(value))
        elif event.key == pygame.K_BACKSPACE:
            self.delete()
        elif event.key in [pygame.K_RETURN, pygame.K_KP_ENTER]:
            self.enter()

    def do_button(self, pos):
        pos = (int((pos[0] - g.DIALPAD[0]) // g.XGAP)
               ), int((pos[1] - g.DIALPAD[1]) // g.XGAP)
        num = 3 * pos[1] + pos[0] + 1

        if pos[1] >= 0 and pos[0] >= 0 and pos[0] < 3:
            if num in range(1, 10):
                if len(self.input) < self.level:
                    self.input.append(num)
            elif num == 10:
                self.enter()
            elif num == 11:
                if len(self.input) < self.level:
                    self.input.append(0)
            elif num == 12:
                self.delete()

    def highlight(self):
        pos = int((g.pos[0] - g.DIALPAD[0]) //
                  g.XGAP), int((g.pos[1] - g.DIALPAD[1]) // g.XGAP)

        if pos[0] in range(3) and pos[1] in range(4):
            num = 3 * pos[1] + pos[0] + 1
            utils.blit_offset('dialpad/highlight', g.DIALPAD, pos)
            utils.blit_offset('dialpad/' + str(num), g.DIALPAD, pos)

    def attempt_disp(self):
        for num in range(len(self.attempts_list)):
            for x in range(self.level):
                utils.blit_offset('status/' + self.compare_list[num][x],
                                  g.ATTEMPT, (self.level + .5 + x, num + 1), 1)
                utils.blit_offset(
                    'entries/' + str(self.attempts_list[num][x]), g.ATTEMPT,
                    (x, num + 1.1), 1)

    def display_answer(self):
        for x in range(self.level):
            utils.blit_offset(
                'entries/' + str(self.num[x]),
                g.ATTEMPT, (x, self.lives + 2), 1)
            utils.blit_offset('status/bull', g.ATTEMPT,
                              (x + 0.5 + self.level, self.lives + 2), 1)

    def enter(self):
        if len(self.input) != self.level:
            pass
        else:
            status = []
            for x in range(self.level):
                if self.num[x] == self.input[x]:
                    status.append('bull')
                elif self.input[x] in self.num:
                    status.append('cow')
                else:
                    status.append('cancel')

            self.compare_list.append(status)
            self.attempts_list.append(self.input)
            if self.input == self.num:
                self.status = 'won'
                self.next_button.set_sensitive(True)
                self.score += (self.lives - len(self.attempts_list)) + 1
                if self.parent:
                    self.parent.update_score(self.score)
            elif len(self.attempts_list) == self.lives:
                self.status = 'lost'
            self.input = []

    def set_next_button(self, next_bt):
        self.next_button = next_bt

    def flush_queue(self):
        flushing = True
        while flushing:
            flushing = False
            if self.journal:
                while Gtk.events_pending():
                    Gtk.main_iteration()
            for event in pygame.event.get():
                flushing = True

    def run(self):
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                return
            elif event.type == pygame.VIDEORESIZE:
                pygame.display.set_mode(
                    (event.size[0], event.size[1] - GRID_CELL_SIZE),
                    pygame.RESIZABLE)
                break
        g.init()
        self.display()
        if self.canvas is not None:
            self.canvas.grab_focus()

        while self.going:
            # Pump Gtk messages.
            if self.journal:
                while Gtk.events_pending():
                    Gtk.main_iteration()
                if not self.going:
                    break

            for event in pygame.event.get():
                if event.type == pygame.MOUSEMOTION:
                    g.pos = event.pos
                    g.redraw = True
                    if self.canvas is not None:
                        self.canvas.grab_focus()

                elif event.type == pygame.MOUSEBUTTONDOWN:
                    if not self.status:
                        self.do_button(event.pos)
                        g.redraw = True
                    self.flush_queue()

                elif event.type == pygame.KEYDOWN:
                    if not self.status:
                        self.do_key(event)
                        g.redraw = True
                    self.flush_queue()

                elif event.type == pygame.QUIT:
                    return

                elif event.type == pygame.VIDEORESIZE:
                    pygame.display.set_mode(
                        (event.size[0], event.size[1] - GRID_CELL_SIZE),
                        pygame.RESIZABLE)
                    g.redraw = True
                    break

            if g.redraw:
                self.display()
                g.screen.blit(g.pointer, g.pos)
                pygame.display.flip()
                g.redraw = False

            pygame.display.update()
            self.clock.tick(20)  # Try to stay at 20 FPS

    def restart(self):
        self.change_level(self.level)

    def change_level(self, level):
        self.level = level
        self.journal = True  # set to False if we come in via main()
        self.num = utils.get_num(self.level)
        self.lives, self.offset = utils.get_lives(self.level)
        self.status = None
        self.input = []
        self.attempts_list = []
        self.compare_list = []
        self.score = 0
        self.parent.update_score(self.score)
        self.next_button.set_sensitive(False)
        g.redraw = True

    def nextRound(self):
        self.next_button.set_sensitive(False)
        self.journal = True  # set to False if we come in via main()
        self.num = utils.get_num(self.level)
        self.lives, self.offset = utils.get_lives(self.level)
        self.status = None
        self.input = []
        self.attempts_list = []
        self.compare_list = []
        g.redraw = True


if __name__ == "__main__":
    pygame.init()
    pygame.display.set_mode((0, 0), pygame.FULLSCREEN)
    game = CowBulls()
    game.journal = False
    game.run()
    pygame.display.quit()
    pygame.quit()
    sys.exit(0)


sugargame/canvas.py
import os
from gi.repository import Gtk
from gi.repository import GObject
from gi.repository import GLib
from sugar3.activity.activity import PREVIEW_SIZE
import pygame
import sugargame.event as event

CANVAS = None


class PygameCanvas(Gtk.EventBox):
    def __init__(self, activity, pointer_hint=True,
                 main=None, modules=[pygame]):
        GObject.GObject.__init__(self)

        global CANVAS
        assert CANVAS is None, "Only one PygameCanvas can be created, ever."
        CANVAS = self

        # Initialize Events translator before widget gets "realized".
        self.translator = event.Translator(activity, self)

        self._activity = activity
        self._main = main
        self._modules = modules

        self.set_can_focus(True)

        self._socket = Gtk.Socket()
        self._socket.connect('realize', self._realize_cb)
        self.add(self._socket)

        self.show_all()

    def _realize_cb(self, widget):

        # Preinitialize Pygame with the X window ID.
        os.environ['SDL_WINDOWID'] = str(widget.get_id())
        for module in self._modules:
            module.init()

        # Restore the default cursor.
        widget.props.window.set_cursor(None)

        # Confine the Pygame surface to the canvas size
        r = self.get_allocation()
        self._screen = pygame.display.set_mode((r.width, r.height),
                                               pygame.RESIZABLE)

        # Hook certain Pygame functions with GTK equivalents.
        self.translator.hook_pygame()

        # Call the caller's main loop as an idle source
        if self._main:
            GLib.idle_add(self._main)

    def get_pygame_widget(self):
        return self._socket

    def get_preview(self):
        """
        Return preview of main surface
        How to use in activity:
            def get_preview(self):
                return self.game_canvas.get_preview()
        """

        if not hasattr(self, '_screen'):
            return None

        _tmp_dir = os.path.join(self._activity.get_activity_root(),
                                'tmp')
        _file_path = os.path.join(_tmp_dir, 'preview.png')

        width = PREVIEW_SIZE[0]
        height = PREVIEW_SIZE[1]
        _surface = pygame.transform.scale(self._screen, (width, height))
        pygame.image.save(_surface, _file_path)

        f = open(_file_path, 'rb')
        preview = f.read()
        f.close()
        os.remove(_file_path)

        return preview


activity.py
# activity.py
# my standard link between sugar and my activity
"""
    Copyright (C) 2018  Rahul Bothra

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

"""

from gettext import gettext as _

import gi
gi.require_version('Gtk', '3.0')
gi.require_version('Gdk', '3.0')

from gi.repository import Gtk
from gi.repository import Gdk
from gi.repository import GdkPixbuf

from sugar3.activity import activity
from sugar3.graphics.toolbarbox import ToolbarBox
from sugar3.activity.widgets import ActivityToolbarButton, StopButton
from sugar3.graphics.combobox import ComboBox
from sugar3.graphics.toolbutton import ToolButton
from sugar3.graphics.style import GRID_CELL_SIZE

import pygame
import sugargame.canvas
import CowBulls


class CowBullsActivtiy(activity.Activity):

    def __init__(self, handle):
        activity.Activity.__init__(self, handle)
        # No sharing (Future Improvement)
        self.max_participants = 1

        # Build the activity toolbar.
        toolbox = ToolbarBox()

        activity_button = ActivityToolbarButton(self)
        toolbox.toolbar.insert(activity_button, 0)
        activity_button.show()

        comboLabel1 = Gtk.ToolItem()
        label1 = Gtk.Label()
        label1.set_text(_('Level:') + ' ')
        comboLabel1.add(label1)
        toolbox.toolbar.insert(comboLabel1, -1)
        comboLabel1.show()
        label1.show()

        comboField = Gtk.ToolItem()
        combo = ComboBox()
        combo.set_wrap_width(3)
        combo.append_item(0, _('Easy'))
        combo.append_item(1, _('Moderate'))
        combo.append_item(2, _('Hard'))
        combo.set_active(0)
        comboField.add(combo)
        combo.connect('changed', self.change_combo)
        toolbox.toolbar.insert(comboField, -1)
        comboField.show()
        combo.show()

        separator = Gtk.SeparatorToolItem()
        separator.props.draw = True
        separator.set_expand(False)
        toolbox.toolbar.insert(separator, -1)
        separator.show()

        restart = ToolButton('new-game')
        toolbox.toolbar.insert(restart, -1)
        restart.set_tooltip(_('Restart'))
        restart.connect('clicked', self._restart_button_cb)
        restart.show()

        next_bt = ToolButton('next')
        toolbox.toolbar.insert(next_bt, -1)
        next_bt.set_tooltip(_('Next Number'))
        next_bt.connect('clicked', self._next_button_cb)
        next_bt.set_sensitive(False)
        next_bt.show()

        separator2 = Gtk.SeparatorToolItem()
        separator2.props.draw = True
        separator2.set_expand(False)
        toolbox.toolbar.insert(separator2, -1)
        separator2.show()

        comboLabel2 = Gtk.ToolItem()
        label1 = Gtk.Label()
        label1.set_text(_('Score:') + '  ')
        comboLabel2.add(label1)
        toolbox.toolbar.insert(comboLabel2, -1)
        comboLabel2.show()
        label1.show()

        self._score_image = Gtk.Image()
        item = Gtk.ToolItem()
        item.add(self._score_image)
        toolbox.toolbar.insert(item, -1)
        item.show()

        separator2 = Gtk.SeparatorToolItem()
        separator2.props.draw = False
        separator2.set_expand(True)
        toolbox.toolbar.insert(separator2, -1)
        separator2.show()

        stop_button = StopButton(self)
        toolbox.toolbar.insert(stop_button, -1)
        stop_button.show()
        stop_button.connect('clicked', self._stop_cb)

        toolbox.show()
        self.set_toolbar_box(toolbox)

        # Create the game instance.
        self.game = CowBulls.CowBulls(parent=self)

        # Build the Pygame canvas.
        self.game.canvas = self._pygamecanvas = sugargame.canvas.PygameCanvas(
            self, main=self.game.run, modules=[pygame.display, pygame.font])

        # Note that set_canvas implicitly calls
        # read_file when resuming from the Journal.
        self.game.set_next_button(next_bt)
        self.set_canvas(self._pygamecanvas)
        Gdk.Screen.get_default().connect('size-changed',
                                         self.__configure_cb)

    def _stop_cb(self, button):
        self.game.going = False

    def change_combo(self, combo):
        level = combo.get_value()
        self.game.change_level(int(level) + 3)

    def get_preview(self):
        return self._pygamecanvas.get_preview()

    def __configure_cb(self, event):
        ''' Screen size has changed '''
        pygame.display.set_mode((Gdk.Screen.width(),
                                 Gdk.Screen.height() - GRID_CELL_SIZE),
                                pygame.RESIZABLE)
        self.game.g_init()
        self.game.run()

    def update_score(self, score):
        pixbuf = self._svg_str_to_pixbuf(self._score_icon(score))
        self._score_image.set_from_pixbuf(pixbuf)
        self._score_image.show()

    def _svg_str_to_pixbuf(self, svg_string):
        ''' Load pixbuf from SVG string '''
        pl = GdkPixbuf.PixbufLoader.new_with_type('svg')
        pl.write(svg_string.encode())
        pl.close()
        pixbuf = pl.get_pixbuf()
        return pixbuf

    def _score_icon(self, score):
        return \
            '<?xml version="1.0" encoding="UTF-8" standalone="no"?>\n' + \
            '<svg\n' + \
            'xmlns:dc="http://purl.org/dc/elements/1.1/"\n' + \
            'xmlns:cc="http://creativecommons.org/ns#"\n' + \
            'xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"\n' + \
            'xmlns:svg="http://www.w3.org/2000/svg"\n' + \
            'xmlns="http://www.w3.org/2000/svg"\n' + \
            'version="1.1"\n' + \
            'width="55"\n' + \
            'height="55"\n' + \
            'viewBox="0 0 55 55">\n' + \
            '<path\n' + \
            'd="M 27.497,50.004 C 39.927,50.004 50,39.937 50,27.508 50,'\
            '15.076 39.927,4.997 27.497,4.997 15.071,4.997 5, \
                 15.076 5,27.508 '\
            '5,39.937 15.071,50.004 27.497,50.004 z"\n' + \
            'style="fill:#ffffff;fill-opacity:1" /><text\n' + \
            'style=" \
                 fill:#000000;fill-opacity:1;stroke:none;font-family:Sans">'\
            '<tspan\n' + \
            'x="27.5"\n' + \
            'y="37.3"\n' + \
            'style="font-size:24px;text-align:center;text-anchor:middle">'\
            '%d' % score + \
            '</tspan></text>\n' + \
            '</svg>'

    def _restart_button_cb(self, event):
        self.game.restart()

    def _next_button_cb(self, event):
        self.game.nextRound()
